src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -139,94 +139,137 @@ abstract class Genotype extends Serializable {
-      val gqq = queryOpt[java.lang.Integer]("GQ")
-      val plq = queryOpt[IndexedSeq[Int]]("PL")
-      val gpq = queryOpt[IndexedSeq[Double]]("GP")
+      def queryInt(field: String): (Row) => Int = {
+        if (s.hasField(field)) {
+          val f = s.field(field)
+          val i = f.index
+          if (f.typ != TInt32)
+            (r: Row) => -1
+          else
+            (r: Row) =>
+              if (r.isNullAt(i))
+                -1
+              else
+                r.getInt(i)
+        } else {
+          (r: Row) => -1
+        }
+      }
 
-      if (plq.isDefined || gpq.isEmpty)
-        (a: Any) =>
-          new RowGenotype(Row(
-            gtq.map(_ (a)).orNull,
-            adq.map(_ (a)).orNull,
-            dpq.map(_ (a)).orNull,
-            gqq.map(_ (a)).orNull,
-            plq.map(_ (a)).orNull,
-            false,
-            false))
-      else
-        (a: Any) => {
-          val gp = gpq.map(_ (a)).orNull
-
-          new RowGenotype(Row(
-            gtq.map(_ (a)).orNull,
-            adq.map(_ (a)).orNull,
-            dpq.map(_ (a)).orNull,
-            gqq.map(_ (a)).orNull,
-            if (gp != null)
-              Genotype.weightsToLinear(gp.toArray)
-            else
-              null,
-            false,
-            true))
+      def queryArrayInt(field: String): (Row) => Array[Int] = {
+        if (s.hasField(field)) {
+          val f = s.field(field)
+          val i = f.index
+          if (f.typ != TArray(TInt32))
+            (r: Row) => null
+          else
+            (r: Row) =>
+              if (r.isNullAt(i))
+                null
+              else {
+                val is = r.getAs[IndexedSeq[Int]](i)
+                val n = is.length
+                val a = new Array[Int](n)
+                var j = 0
+                while (j < n) {
+                  a(j) = is(j)
+                  j += 1
+                }
+                a
+              }
+        } else {
+          (r: Row) => null
+        }
+      }
+
+      def queryArrayDouble(field: String): (Row) => Array[Double] = {
+        if (s.hasField(field)) {
+          val f = s.field(field)
+          val i = f.index
+          if (f.typ != TArray(TFloat64))
+            (r: Row) => null
+          else
+            (r: Row) => {
+              if (r.isNullAt(i))
+                null
+              else {
+                val is = r.getAs[IndexedSeq[Double]](i)
+                val n = is.length
+                val a = new Array[Double](n)
+                var j = 0
+                while (j < n) {
+                  a(j) = is(j)
+                  j += 1
+                }
+                a
+              }
+            }
+        } else {
+          (r: Row) => null
         }
+      }
+
+      val gtq = queryCall("GT")
+      val adq = queryArrayInt("AD")
+      val dpq = queryInt("DP")
+      val gqq = queryInt("GQ")
+
+      if (s.hasField("PL") || !s.hasField("GT")) {
```

Comment:
I'm uncomfortable with this logic. I realize that this was already in Hail in some form, but it makes me squirm a bit.